### PR TITLE
Add version retention period to Spanner database resource

### PR DESCRIPTION
**Pull request by SarahFrench**
# Description

Fixes https://github.com/hashicorp/terraform-provider-google/issues/8994

This PR lets Terraform users set the retention period of historical versions of a Spanner database's data and schema for [point in time recovery (PITR)](https://cloud.google.com/spanner/docs/pitr).

The retention period of the data used for PITR can be changed by running a DDL statement to update the database's options:

```sql
-- Google Standard SQL
ALTER DATABASE `example-db` SET OPTIONS (version_retention_period="7d";

-- Postgres
ALTER DATABASE `example-db` SET spanner.version_retention_period TO "7d";
```

[See usage notes section here](https://cloud.google.com/spanner/docs/use-pitr) & [docs about equivalent DDL for Postgres](https://cloud.google.com/spanner/docs/reference/postgresql/data-definition-language#alter-database)

These DDL statements can be sent through the REST API and executed at either:
1. **Creation of a new database via the `create` endpoint**
    - as an entry in the `extraStatements` array [in the request body](https://cloud.google.com/spanner/docs/reference/rest/v1/projects.instances.databases/create#request-body)
    - ⚠️ **not allowed for Postgres** ⚠️
        - If you provide any `extraStatements` in the create request for a Postgres database, you get this error: `DDL statements other than <CREATE DATABASE> are not allowed in database creation request for PostgreSQL-enabled databases`
2. **Updating a database via the [`updateDdl`](https://cloud.google.com/spanner/docs/reference/rest/v1/projects.instances.databases/updateDdl) endpoint**
    - OK for both database dialects

## Approach

I've used the value of the `version_retention_period` field to create an ALTER DDL statement like above and add it to the array of statements in the request bodies of requests sent to the `create` or `updateDdl` endpoints

**Edit 2022-06-27** : I refactored the code for creating databases so that DDL statements (both originating from `ddl` and `version_retention_period` fields) are executed in the database via a separate API call. After the database is made using the `create` endpoint the provider runs DDL against the new database using the `updateDdl` endpoint, all within the creation process in Terraform.

## Postgres issues

~~The DDL restriction when making Postgres databases has been met by the initial version of this PR by skipping setting the retention time during creation, even if the `version_retention_period` is set in the Terraform configuration. When a plan is generated a 2nd time from the Terraform configuration it will plan to update the retention time on the Postgres DB.~~

~~This might be confusing to practitioners, but seems like the need to re-apply plans for Postgres databases [isn't new and is mentioned in the docs](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/spanner_database#database_dialect:~:text=You%20must%20therefore%20re%2Dapply%20terraform%20with%20ddl%20on%20the%20same%20database%20after%20creation.%20Possible%20values%20are%20GOOGLE_STANDARD_SQL%20and%20POSTGRESQL.).~~

**Edit 2022-06-27** : By separating all execution of DDL into an API call separate to the initial `create` API call this means Postgres's limitations in Spanner don't require 2nd apply steps

# Open questions

- How does this approach work with acceptance tests, which try to prevent scenarios where 2nd plans from the same config contain changes still. Is that a sign that this should be approached differently?
- Would a more explicit error that disallows `version_retention_period` being set before making the Postgres DB be better? That would avoid the confusion of needing to apply a plan twice to reach the desired goal
- Should there be any protection against users setting the retention time via an `ALTER` statement in the `ddl` parameter of the Terraform provider? The only issue would be if someone added `version_retention_period` to their config but then resumed setting it via `ddl` (`version_retention_period` would always re-assert itself on the next plan&apply)

# PR steps

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
spanner: Added field `version_retention_period` to `google_spanner_database` resource
```
```release-note:enhancement
spanner: Fixed issue where `ddl` and `version_retention_period` could not be set when first creating `google_spanner_database` using POSTGRESQL
```

### CODE REVIEW

**Comment by SarahFrench** [comment]
Squashed a couple of my commits together to clean up the PR and accidentally removed the link between commits & modular-magician output - sorry about that! 😅 

When I first made the PR I wasn't sure to handle how the new field cannot be set when creating a new Postgres DB. In this PR I've included ways that side-step this restriction in the API, but as the same isn't done for `ddl` and the API error is returned. Maybe I should simplify this PR to follow suit? My motivation for doing it in a different way was that the API error is easily connected to the `ddl` property as it mentions DDL, but some may not realise that the new field is just another DDL statement under the hood

**Review by c2thorn — CHANGES_REQUESTED** [review]
only minor asks, looks great overall. I'll address the open questions in another comment

**Comment by c2thorn** [comment]
> How does this approach work with acceptance tests, which try to prevent scenarios where 2nd plans from the same config contain changes still. 

We would ignore the field when verifying the create step like here: https://github.com/GoogleCloudPlatform/magic-modules/pull/5799/files#diff-bc14ae2ed214f27590f8fe2a8106235c9dee0fca59f31608041d97e025b257c8R143
Not needed in your test since it is going to the default during create and updating later.

> Is that a sign that this should be approached differently?

Yep, but isn't too rare. Also see https://github.com/GoogleCloudPlatform/magic-modules/pull/6141#discussion_r903099676

> Would a more explicit error that disallows version_retention_period being set before making the Postgres DB be better? That would avoid the confusion of needing to apply a plan twice to reach the desired goal

It's hard for me to tell if it's better to try to warn users up front vs the convenience of not having to edit the config twice. Many users run Terraform effectively in a `terraform apply` loop (against recommendation of course) which the error would hinder. If anything, I defer to the precedent set by https://github.com/GoogleCloudPlatform/magic-modules/pull/5799

> Should there be any protection against users setting the retention time via an ALTER statement in the ddl parameter of the Terraform provider? The only issue would be if someone added version_retention_period to their config but then resumed setting it via ddl (version_retention_period would always re-assert itself on the next plan&apply)

That feels like a minor enough use case to not warrant the effort + maintenance, but let me know if you think otherwise. If they wanted to resume setting it via `ddl`, they would just need to remove the `version_retention_period` field right?

File: mmv1/templates/terraform/constants/spanner_database.go.erb [hunk 1]
```diff
@@ -42,4 +42,42 @@ func resourceSpannerDBDdlCustomDiffFunc(diff TerraformResourceDiff) error {
 func resourceSpannerDBDdlCustomDiff(_ context.Context, diff *schema.ResourceDiff, meta interface{}) error {
 	// separate func to allow unit testing
 	return resourceSpannerDBDdlCustomDiffFunc(diff)
+}
+
+func validateDatabaseRetentionPeriod(v interface{}, k string) (ws []string, errors []error) {
```
[the remark this answers is not in the document]
Comment (SarahFrench) [reply mmv1/templates/terraform/constants/spanner_database.go.erb:47]: That makes sense, I'll avoid duplicating logic in the provider in future- thanks for the feedback! I'll leave this code in for now but I'm happy to remove it too.

**Comment by SarahFrench** [comment]
Hi @c2thorn, thanks for your review. After I learned about the ability to add post_create blocks I changed approach for this PR - [covered here in this comment](https://github.com/GoogleCloudPlatform/magic-modules/pull/6141#discussion_r907684559) - which goes against the precedent set by https://github.com/GoogleCloudPlatform/magic-modules/pull/5799 but I agree results in better UX. I believe I've updated the necessary tests & documentation to describe this and no longer warn about needing to apply the config 2 times.

Also, re this from your last [comment](https://github.com/GoogleCloudPlatform/magic-modules/pull/6141#issuecomment-1162421791):

>If they wanted to resume setting it via ddl, they would just need to remove the version_retention_period field right?

I added a new acceptance test in commit e302e6d that documents the expected behaviour of how the two fields interact - it's possible to swap back and forth between the two without any major downsides (other than out-of-date SQL in `ddl` that doesn't reflect the current state of the DB).

**Pull request merged**